src/vs/platform/terminal/node/terminalEnvironment.ts

```diff
@@ -188,7 +188,10 @@ export function getShellIntegrationInjection(
 			const zdotdir = path.join(os.tmpdir(), `${os.userInfo().username}-vscode-zsh`);
 			envMixin['ZDOTDIR'] = zdotdir;
 			const userZdotdir = env?.ZDOTDIR ?? os.homedir() ?? `~`;
-			envMixin['USER_ZDOTDIR'] = userZdotdir;
+			// NOTE: When vscode is started from the integrated terminal, the environment variables are taken over.
+			// The ZDOTDIR variable is oriented to os.tmpdir(), so we cannot reuse.
+			// Therefore, we use USER_ZDOTDIR variable that is inherited.
+			envMixin['USER_ZDOTDIR'] = env?.USER_ZDOTDIR ?? userZdotdir;
```

Comment:
thanks for this explanation - it makes a lot of sense. 

While this change would fix it, I think perhaps better would be to remove this line

https://github.com/microsoft/vscode/blob/3161548771dc0e01556e7116bbcd57e0aaa32a26/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-env.zsh#L12

Reply:
By deleting that one line, it is no longer necessary to modify `src/vs/platform/terminal/node/terminalEnvironment.ts` !
It makes more sense that way.

Reply:
Oops, but if I incorporate that change, command actions, etc. won't work...

Reply:
Then I think it would be better to revert these two commits, what do you think?
https://github.com/microsoft/vscode/pull/161433/commits/4c6670df6fb68e10267406a66c171b4d99cde190
https://github.com/microsoft/vscode/pull/161433/commits/bcfad3268a11e807f4a611e6f4e64fbb929d3abe

In short, it is the state of the code at the phase you have reviewed.

Reply:
> While this change would fix it, I think perhaps better would be to remove this line, which sets the `ZDOTDIR` back to the vscode one unnecessarily
> 
> https://github.com/microsoft/vscode/blob/3161548771dc0e01556e7116bbcd57e0aaa32a26/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-env.zsh#L12

@meganrogge I think the model was when we enter and exit any zsh init script we want our zdotdir to be set to `$VSCODE_ZDOTDIR`, such that our login, rc, etc. correctly get run. With the exception of rc where we do not which is what this was about:

https://github.com/microsoft/vscode/blob/430f212a16b1575e10a562b709f52399869401d7/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-rc.zsh#L131-L133

Should we be doing that in login shells as well? It appears `$ZDOTDIR` remains the vscode path for login shells:

https://github.com/microsoft/vscode/blob/430f212a16b1575e10a562b709f52399869401d7/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-rc.zsh#L19-L24

Could that instead be this?

```ts
if [[ $options[norcs] = off  && -f $USER_ZDOTDIR/.zshrc ]]; then 
 	VSCODE_ZDOTDIR=$ZDOTDIR 
 	ZDOTDIR=$USER_ZDOTDIR 
 	. $USER_ZDOTDIR/.zshrc
 fi 
```

And then remove this:

https://github.com/microsoft/vscode/blob/430f212a16b1575e10a562b709f52399869401d7/src/vs/workbench/contrib/terminal/browser/media/shellIntegration-rc.zsh#L131-L133

Since `$USER_ZDOTDIR` is not actually set when `$VSCODE_INJECTION !== 1`

Reply:
Yes I had reversed the order of the startup files and was thinking `env` runs last, but you're correct

Reply:
and we might also want to add a `shellIntegration-logout.zsh` file ?

Reply:
![Screen Shot 2022-09-23 at 1 02 08 PM](https://user-images.githubusercontent.com/29464607/192048078-3bf323c6-9f18-4ef5-bcf8-0461717f222e.png)

Reply:
On logout I think we don't need to run anything special so if it was the regular `$ZDOTDIR` at that point it would work fine

Reply:
yeah but if we fix it in the `zshrc`, then `.zlogin` won't get run

Reply:
so I believe we need to set it back in `shellIntegration-login.zsh`

Reply:
Yes you're right, whatever the last thing is we want to set it to the user's ZDOTDIR there (H for login, D for non-login)

Reply:
so confusing 😅 thus my bad suggestion at the start of this thread

Reply:
@yamachu sorry for the confusion. pls let us know if you'd like us to take it from here